pkg/live/planner/cluster_test.go

```diff
@@ -0,0 +1,172 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package planner
+
+import (
+	"context"
+	"testing"
+
+	planv1alpha1 "github.com/GoogleContainerTools/kpt/pkg/api/plan/v1alpha1"
+	"github.com/GoogleContainerTools/kpt/thirdparty/cli-utils/apply"
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cli-utils/pkg/apply/event"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/testutil"
+)
+
+var (
+	deploymentYAML = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: foo
+  namespace: default
+spec:
+  replicas: 1
+`
+)
+
+func TestClusterPlanner(t *testing.T) {
+	testCases := map[string]struct {
+		resources        []*unstructured.Unstructured
+		clusterResources []*unstructured.Unstructured
+		events           []event.Event
+
+		expectedPlan *planv1alpha1.Plan
+	}{
+		"single new resource": {
+			resources: []*unstructured.Unstructured{
+				testutil.Unstructured(t, deploymentYAML),
+			},
+			clusterResources: []*unstructured.Unstructured{},
+			events: []event.Event{
+				{
+					Type: event.InitType,
+					InitEvent: event.InitEvent{
+						ActionGroups: event.ActionGroupList{
+							{
+								Action: event.ApplyAction,
+								Name:   "apply-1",
+								Identifiers: []object.ObjMetadata{
+									testutil.ToIdentifier(t, deploymentYAML),
+								},
+							},
+						},
+					},
+				},
+				{
+					Type: event.ApplyType,
+					ApplyEvent: event.ApplyEvent{
+						GroupName:  "apply-1",
+						Identifier: testutil.ToIdentifier(t, deploymentYAML),
+						Operation:  event.ServersideApplied,
+						Resource:   testutil.Unstructured(t, deploymentYAML),
+					},
+				},
+			},
+			expectedPlan: &planv1alpha1.Plan{
+				ResourceMeta: planv1alpha1.ResourceMeta,
+				Spec: planv1alpha1.PlanSpec{
+					Actions: []planv1alpha1.Action{
+						{
+							Type:      planv1alpha1.Create,
+							Name:      "foo",
+							Namespace: "default",
+							Group:     "apps",
+							Kind:      "Deployment",
+							After:     testutil.Unstructured(t, deploymentYAML),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for tn := range testCases {
+		tc := testCases[tn]
+		t.Run(tn, func(t *testing.T) {
+			ctx := context.Background()
+
+			applier := &FakeApplier{
+				events: tc.events,
+			}
+
+			fakeResourceFetcher := &FakeResourceFetcher{
+				resources: tc.clusterResources,
+			}
+
+			plan, err := (&ClusterPlanner{
+				applier:         applier,
+				resourceFetcher: fakeResourceFetcher,
+			}).BuildPlan(ctx, &FakeInventoryInfo{}, []*unstructured.Unstructured{}, Options{})
+			require.NoError(t, err)
+
+			if diff := cmp.Diff(tc.expectedPlan, plan); diff != "" {
+				t.Errorf("plan mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+type FakeApplier struct {
+	events []event.Event
```

Comment:
Observation: From a testing perspective, it would be great if we didn't have to construct the events, but could instead provide the pre-existing resources.  Not for this PR though!

Reply:
Agree.